nova/compute/api.py

```diff
@@ -695,6 +695,33 @@ def _merge_bdms_lists(self, overridable_mappings, overrider_mappings):
                 [bdm for bdm in overridable_mappings
                  if bdm['device_name'] not in device_names])
 
+    def _flavor_enforces_boot_from_volume(self, flavor):
+        """Determines whether boot-from-volume is enforced by flavor"""
+        return (flavor.get('extra_specs', {})
+                .get('boot_from_volume', 'false') == 'true')
+
+    def _enforce_boot_from_volume(self, image_meta, flavor,
+                                  block_device_mapping):
+        """Adds boot-from-volume block device mapping if not already set."""
+        root_bdm = block_device.get_root_bdm(block_device_mapping)
+        if root_bdm:
+            if root_bdm.get('destination_type') == 'volume':
+                # User already set a root volume to boot from.
+                return
+            elif (root_bdm.get('destination_type') == 'local' and
+                  root_bdm.get('source_type') == 'image'):
```

Comment:
I'd switch the order here. Source -> Destination.

Or even refactor out into a separate variable or function that describes what this combination of conditions actually means. I read [in the docs](https://docs.openstack.org/nova/latest/user/block-device-mapping.html#valid-source-destination-combinations) that _image_ -> _local_ is special for booting directly from a Glance image(?), and this could be made more explicit here.